### ISSUE TITLE
gvim can't map <S-Esc> separately from <Esc>

### ISSUE THREAD

**Issue by RedBeard0531** [vim/vim#2520]
In mintty I like using <S-Esc> (mapped from <9b>) to exit terminal mode, and I'd like to be able to do the same in gvim. It seems like gvim treats <S-Esc> the same as <Esc> so this isn't possible today.

**Comment by mattn** [comment]
What OS do you use?

**Comment by RedBeard0531** [comment]
I use gvim on archlinux. I compile it myself to use gtk2 (due to #1199) and enable terminal support.

**Comment by k-takata** [comment]
Perhaps, this is a GTK specific issue?
Windows GUI version can map `<S-Esc>` separately from `<Esc>`.

**Comment by nuko8** [comment]
Sorry, I already tried reproducing the issue a few days ago and, as what I got was a bit different from what the OP reported (at least in appearance), I had been in wait-and-see mode since then (Actually, I've been busy  recently :)

What I got was as follows: For all the GUIs running on X11, namely, Athena, Motif, GTK+ 2 and 3, they couldn't distinguish `<S-Esc>` from `<Esc>` when they were launched with `gvim -u NONE`, but they did with `gvim -u NONE -N`.

Interestingly, MacVim showed the same behavior (`gvim` was replaced with `mvim` when I launched it from the command line).

Summing up, I concluded that, with non-compatible mode at least, all the GUIs for UNIX-like systems are able to  "map `<S-Esc>` separately from `<Esc>`".

So I cannot be sure if the issue is a bug of Vim or not.

**Comment by vim-ml** [comment]
On Mon, Jan 8, 2018 at 7:42 AM, Kazunobu Kuriyama
<vim-dev-github@256bit.org> wrote:
> Sorry, I already tried reproducing the issue a few days ago and, as what I
> got was a bit different from what the OP reported (at least in appearance),
> I had been in wait-and-see mode since then (Actually, I've been busy
> recently :)
>
> What I got was as follows: For all the GUIs running on X11, namely, Athena,
> Motif, GTK+ 2 and 3, they couldn't distinguish <S-Esc> from <Esc> when they
> were launched with gvim -u NONE, but they did with gvim -u NONE -N.
>
> Interestingly, MacVim showed the same behavior (gvim was replaced with mvim
> when I launched it from the command line).
>
> Summing up, I concluded that, with non-compatible mode at least, all the
> GUIs for UNIX-like systems are able to "map <S-Esc> separately from <Esc>".
>
> So I cannot be sure if the issue is a bug of Vim or not.

When I start gvim (Huge version 8.0.1428 with GTK2-GNOME GUI) with -u
NONE -U (and the package xorg-x11-server being at version
7.6_1.18.3-28.1, and libxcb1 and libxcb-devel at version 1.11.1-6.1,
as distributed as part of openSUSE Leap 42.3 with all updates
applied), then start Insert mode and hit Ctrl-V followed by Shift-Esc,
what is inserted (as seen by going back to Normal mode and hitting ga)
is a plain Esc but this is not necessarily a bug.

Vim intentionally does not distinguish between Ctrl-a and
Ctrl-Shift-a, Ctrl-b and Ctrl-Shift-b, etc., and (Ctrl + printable
key) where the "printable key" is other than A-Z, a-z, @, [, \, ], ^,
_, ? usually either gives nothing or ignores the Ctrl modifier. This
is a legacy from ASCII where those combinations were used for 0x00 to
0x3F plus 0x7F (a total of 33 values i.e. 26+7) and where the
difference between upper- and lower-case was intentionaly neutralized
in the presence of Ctrl. Legacy vi followed ASCII in that regard and
you know (or should know) how Bram is attached to upward
compatibility.

In Console Vim the results are even less reproducible since in that
case there is an additional layer (the console or console emulator)
between the keyboard and Vim.


Best regards,
Tony.

**Comment by RedBeard0531** [comment]
Hmm, I just tried `tnoremap <silent><S-Esc> <C-\><C-n>` and it worked fine. I gave up too early because my normal way of mapping special keys is typing ^V<KEY> and seeing what gets output. Since it was still outputting ^] I just assumed it wouldn't work.

Sorry about the false report. Closing.

**Comment by tonymec** [comment]
On Mon, Jan 8, 2018 at 5:49 PM, Mathias Stearn <vim-dev-github@256bit.org> wrote:
> Hmm, I just tried `tnoremap <silent><S-Esc> <C-\><C-n>` and it worked fine. I
> gave up too early because my normal way of mapping special keys is typing ^V
> and seeing what gets output. Since it was still outputting ^] I just assumed
> it wouldn't work.

Hm, so do I, usually (see above). But your last comment made me think. So I did
```
:new
:imap <buffer> <S-Esc> +++
```
then went into Insert mode, hit `Shift-Esc`, and lo! `+++` was inserted.
So my gvim can tell that I hit `<S-Esc>` after all, it's just that if that key is not remapped it will insert a plain `<Esc>`

I wouldn't call it a "false" report, maybe at worst an insufficiently researched one, and it taught me something — thank you.
